onnxruntime/python/tools/transformers/gpt2_beamsearch_helper.py

```diff
@@ -0,0 +1,822 @@
+            output_unfinished_sents,
+            current_step_results.view(self.config.batch_size * self.config.beam_size, -1),
+            current_step_scores.view(self.config.batch_size * self.config.beam_size, -1),
+        )
+
+
+# Maps model class name to a tuple of model class, name of first output and use padding or not
+MODEL_CLASSES = {
+    'GPT2LMHeadModel': (MyGPT2LMHeadModel, 'logits', True),
+    'GPT2LMHeadModel_NoPadding': (MyGPT2LMHeadModel_NoPadding, 'logits', False),
+    'GPT2Model': (MyGPT2Model, 'last_state', True),
+    "GPT2LMHeadModel_BeamSearchStep": (GPT2LMHeadModel_BeamSearchStep, "last_state", True), # defined in gpt2_beamsearch_helper.py
+}
+
+
+class Gpt2BeamSearchInputs(Gpt2Inputs):
+    def __init__(
+        self,
+        input_ids,
+        position_ids,
+        attention_mask,
+        past,
+        beam_select_idx=None,
+        input_log_probs=None,
+        input_unfinished_sents=None,
+        prev_step_results=None,
+        prev_step_scores=None,
+    ):
+        super().__init__(input_ids, position_ids, attention_mask, past)
+        self.prev_step_results: torch.LongTensor = prev_step_results
+        self.prev_step_scores: torch.FloatTensor = prev_step_scores
+        if beam_select_idx is None:
+            self.beam_select_idx: torch.LongTensor = torch.zeros(
+                [1, len(input_ids)]
+            ).long()
+        else:
+            self.beam_select_idx: torch.LongTensor = beam_select_idx
+        self.input_log_probs: torch.FloatTensor = input_log_probs
+        self.input_unfinished_sents: torch.ByteTensor = input_unfinished_sents
+
+    def to_list(self) -> List:
+        input_list = [
+            v 
+            for v in [
+                self.input_ids,
+                self.position_ids,
+                self.attention_mask,
+                self.beam_select_idx, 
+                self.input_log_probs, 
+                self.input_unfinished_sents, 
+                self.prev_step_results, 
+                self.prev_step_scores
+            ]
+            if v is not None
+        ]
+        if self.past:
+            input_list.extend(self.past)
+        return input_list
+
+    def to_tuple(self) -> Tuple:
+        return tuple(
+            v
+            for v in [
+                self.input_ids,
+                self.position_ids,
+                self.attention_mask,
+                self.past,
+                self.beam_select_idx,
+                self.input_log_probs,
+                self.input_unfinished_sents,
+                self.prev_step_results,
+                self.prev_step_scores,
+            ]
+            if v is not None
+        )
+
+    def to_fp32(self):
+        gpt2_inputs = super().to_fp32()
+        return Gpt2BeamSearchInputs(
+            gpt2_inputs.input_ids,
+            gpt2_inputs.position_ids,
+            gpt2_inputs.attention_mask,
+            gpt2_inputs.past,
+            self.beam_select_idx,
+            self.input_log_probs,
```

Comment:
Shall input_log_probs etc be converted to fp32 (if it could be fp16)?

Reply:
Actually input_log_probs etc can't be fp16

Reply:
> Shall input_log_probs etc be converted to fp32 (if it could be fp16)?

Sorry I misunderstood. Good catch! I have made changes.